### ISSUE TITLE
Validators.compose returns null instead of function when validators are not provided.

### ISSUE THREAD

**Issue by wawyed** [angular/angular#15624]
**I'm submitting a ...**
```
[ ] bug report => search github for a similar issue or PR before submitting
[X] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Currently when invoking Validators.compose with an empty array/undefined parameter it returns null, having to add boilerplate to the code in order to compensate for it.

Example

```javascript  
protected _validatorFn(control: FormControl): Dict<any> {
    let validators: Array<ValidatorFn> = [];

    if (this.isRequired()) {
      validators.push(Validators.required);
    }

    return validators.length ? Validators.compose(validators)(control) : null;
  }
```

**Expected behavior**
Validators.compose should return a function that returns null in this case.

So this can work without having to worry about the internals of compose:

```javascript  
protected _validatorFn(control: FormControl): Dict<any> {
    let validators: Array<ValidatorFn> = [];

    if (this.isRequired()) {
      validators.push(Validators.required);
    }

    return Validators.compose(validators)(control);
  }
```

**Minimal reproduction of the problem with instructions**
Examples provided should suffice to see the issue.

**What is the motivation / use case for changing the behavior?**
Cleaner and more maintainable code for the consumer of Validators.compose

**Please tell us about your environment:**
N/A

* **Angular version:** 
4.0.0

* **Browser:** all 
 
* **Language:** all

* **Node (for AoT issues):** `node --version` =  N/A

**Comment by Toxicable** [comment]
@wawyed why do you need `Validators.compose()` to return something that's not `null`?
The FormControl can take `null` as a parameter for validators and handle it fine so I don't see why null is causing you an issue here.
As for what it might return if it wasn't `null`, I was thinking maybe something like `() => null`,  a validator that is always valid, but that feels super janky to me, therefore I think `null` is the appropriate value here

cc @kara

**Comment by wawyed** [comment]
I need it for the reasons stated above. I am passing a custom validator function that invokes a validator.compose instead since my validators are applied conditionally so I can't pass to FormControl the result of the compose function since I need to generate it dynamically every time the validators are ran.

You really think returning null is the right thing? because I disagree completely. the interface for Validators.compose() is

``` javascript
 compose(validators: ValidatorFn[]): ValidatorFn
```

being validatorFn

``` javascript
export interface ValidatorFn { (c: AbstractControl): ValidationErrors|null; }
```

No where in the definition of it says that is going to return null instead of a function and to be honest, if it did, it would be quite confusing since you would need to specify it via documentation that null will be return in X circumstances. Instead, if you just choose to return () => null. You don't need to document anything extra and it keeps in line with the definition of a validator perfectly, since a validator function can return null.

**Comment by Toxicable** [comment]
@wawyed while you're right that signiture is slightly off, unless i'm mistaken it should be for strictNullChecks
```
 compose(validators: ValidatorFn[]): ValidatorFn|null
```
However i'm not sure what you mean by 
>validators are applied conditionally so I can't pass to FormControl 
Since you can pass `null` to a `FormControl` see here https://plnkr.co/edit/t7ci4a6Fc6FjvLlGWNSn?p=preview

**Comment by wawyed** [comment]
What I mean is that dynamically (For example based on another input value) I want to apply certain validators to my control, so the way that I'm achieving this is by creating a custom validator function that invokes and returns the Validators.compose method. Sort of like this:

``` javascript
let myVal = 3;

function isRequired() {
  return myVal === 3;

}

function _validatorFn(control: FormControl): Dict<any> {
    let validators: Array<ValidatorFn> = [];

    if (isRequired()) {
      validators.push(Validators.required);
    }

    return validators.length ? Validators.compose(validators)(control) : null;
  }

let myControl = new FormControl('initialValue', _validatorFn); 

myControl.setValue(''); //Control is valid.

myVal = 4;

myControl.updateValueAndValidity() //Control is invalid.

```

**Comment by Toxicable** [comment]
What's wrong with using `FormControl#setValidators()`

**Comment by trotyl** [comment]
The signature has been fixed by #15859 in `4.1.0-rc.0`, this is the expected behavior now.

**Comment by AndrewKushnir** [comment]
Hi @wawyed, thanks for creating this ticket. As it was mentioned in this thread, the `Validators.compose` function types were updated to take `null` into account and `null` can be used in other validator-related APIs (like `FormControl#setValidators()`). The Forms module logic seems to work as expected and I'm closing this ticket for now. Thank you.
